Add flux normaliser processor

hrobarts opened the pull request:
## Description
New FluxNormaliser processor to normalise data
- [ ] to a float value
- [ ] to an array of floats with size matching the 
- [ ] to the mean value calculated from a specified region of interest in the data itself

## Example Usage
```
processor = FluxNormaliser(flux=10)
processor.set_input(self.data_cone)
data_norm = processor.get_output()
```

## Changes


## Testing you performed
> Please add any demo scripts to https://github.com/TomographicImaging/CIL-Demos/tree/main/misc
Tests in test_DataProcessor.py TestFluxNormaliser class

## Related issues/links


## Checklist

- [ ] I have performed a self-review of my code
- [ ] I have added docstrings in line with the guidance in the developer guide
- [ ] I have updated the relevant documentation
- [ ] I have implemented unit tests that cover any new or modified functionality
- [ ] CHANGELOG.md has been updated with any functionality change
- [ ] Request review from all relevant developers
- [ ] Change pull request label to 'Waiting for review'

## Contribution Notes

Please read and adhere to the [developer guide](https://tomographicimaging.github.io/CIL/nightly/developer_guide/) and local patterns and conventions.

- [ ] The content of this Pull Request (the Contribution) is intentionally submitted for inclusion in CIL (the Work) under the terms and conditions of the [Apache-2.0](https://www.apache.org/licenses/LICENSE-2.0) License
- [ ] I confirm that the contribution does not violate any intellectual property rights of third parties

hrobarts commented:
Some examples:
Normalise all data with a scalar value
```
processor = FluxNormaliser(flux=10)
processor.set_input(data)
data_norm = processor.get_output()
show2D([data, data_norm], ['Data','Normalised data'])
```
![image](https://github.com/user-attachments/assets/b3e9ff40-8ba9-4041-bf72-b39b1c34e8ce)

Normalise data with a scalar value per projection
```
processor = FluxNormaliser(flux=np.arange(1,2,(2-1)/(data.get_dimension_size('angle'))))
processor.set_input(data)
data_norm = processor.get_output()
show2D([data, data_norm], slice_list=('vertical',80))
```
![image](https://github.com/user-attachments/assets/02443488-c01f-4eeb-80d2-1b49831737eb)

Normalise data using a region in the background of each projection.
```
roi = {'vertical':(0,14), 'horizontal':(0,14)}
processor = FluxNormaliser(roi=roi)
processor.set_input(data)
processor.show_roi()
```
The default output of `show_roi()` shows the roi plotted on the dataset for the first and last projection and the projection which has the maximum variation from the mean intensity (Z-score) in the roi, as well as a plot of the mean intensity per projection. 
![image](https://github.com/user-attachments/assets/4a305734-edb6-4f58-834f-715119cf04b8)
In this case we can see that although the roi is empty for most projections, there are some cases where there is material in the roi like index 176.
We can also use `processor.show_roi(angle_index=250)` with a specific angle_index if we want to check a different projection which looks like an outlier
![image](https://github.com/user-attachments/assets/3572a03f-43cb-4ed2-ad99-ed8fe74eba84)
In this case, if we change the roi we can make sure it's empty for all projections
```
roi = {'vertical':(0,10), 'horizontal':(0,10)}
processor = FluxNormaliser(roi=roi)
processor.set_input(data)
processor.show_roi()
```
![image](https://github.com/user-attachments/assets/308d945c-8dc5-4c9e-bd94-b4fb53fa1bbd)
```
data_norm = processor.get_output()
```
In this example with synchrotron data, the background has a gradient so the Z-score might not be the best metric to use.
```
roi = {'vertical':(10,135), 'horizontal':(125,150)}
processor = FluxNormaliser(roi=roi)
processor.set_input(data)
processor.show_roi()
```
![image](https://github.com/user-attachments/assets/adcd2dec-f7f4-41bd-ac9a-ae996becf1e2)

```
roi = {'vertical':(10,135), 'horizontal':(135,150)}
processor = FluxNormaliser(roi=roi)
processor.set_input(data)
processor.show_roi()
```
![image](https://github.com/user-attachments/assets/ddb80fd8-2a69-4430-89c4-002223eace70)


```
data_norm = processor.get_output()
show2D([data, data_norm], slice_list=('vertical',80))
```
![image](https://github.com/user-attachments/assets/2a8ea7f1-e414-4332-88e2-3db2344d7230)

The gradient is opposite if we use the roi on the other side
```
roi = {'vertical':(10,135), 'horizontal':(10,32)}
processor = FluxNormaliser(roi=roi)
processor.set_input(data)
processor.show_roi()
```
![image](https://github.com/user-attachments/assets/3cd7c0e1-de78-4b2c-916a-931bf4abd244)
```
data_norm = processor.get_output()
show2D([data, data_norm], slice_list=('vertical',80))
```
![image](https://github.com/user-attachments/assets/82dca751-7803-441f-b7e4-01fe0a022b95)

hrobarts commented:
Updated to plot the mean, minimum and maximum in the roi
![image](https://github.com/user-attachments/assets/75ac1fdc-794a-4655-9435-a55d238684a0)

hrobarts commented:
Updated so `preview_configuration()` still works on data with one angle
![image](https://github.com/user-attachments/assets/738eecac-b19a-4ac0-bf6f-ea096a8f589c)

Also works on 2D data where the roi is specified with one of horizontal or vertical. Plot the roi on the sinogram
![image](https://github.com/user-attachments/assets/f659da41-c93c-4550-a679-ec296494af81)

Also works if the dataorder is different
![image](https://github.com/user-attachments/assets/059a8b4c-0e3f-49d6-ab73-0e4ef97db9ed)


Also works if the data has multiple channels, the flux to normalise to is the mean in the roi averaged over the channels. `preview_configuration()` shows the central channel by default
![image](https://github.com/user-attachments/assets/8fadd840-5dde-4e99-a93b-04975b91f3cc)
or the channel can be specified `processor.preview_configuration(channel=0)`
![image](https://github.com/user-attachments/assets/c4d7007a-9cdb-4ff7-94a8-99cf88b95fe6)

hrobarts commented:
@gfardell we discussed whether the preview_configuration should just be an option in logging. At the moment I've left it as a function the users can access - what do you think?

paskino commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,314 @@
+    roi: dict (optional)
+        Dictionary describing the region of interest containing the background
+        in the image. The image is divided by the mean value in the roi. If None,
+        specify flux direcrtly, default is None.
```
```suggestion
        specify flux directly, default is None.
```

paskino commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,314 @@
+        The value to multiply the image by. If None, the mean flux value will 
+        be used, default is None.
+        
+    tolerance: float (optional)
```
is tolerance the right term?

hrobarts replied:
Hi both, I've removed this argument and instead raise an error if the flux contains a 0.

paskino commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,314 @@
+                raise ValueError("Please specify either flux or roi, not both")
+
+            else:
+                if isinstance(self.roi,dict):
```
```suggestion
                if isinstance(self.roi, dict):
```

lauramurgatroyd commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,314 @@
+        specify flux direcrtly, default is None.
+
+    norm_value: float (optional)
+        The value to multiply the image by. If None, the mean flux value will 
+        be used, default is None.
```
'mean flux' could be defined a bit more clearly. For the case where we don't provide flux and instead provide an ROI, is this the mean of the ROI across all projections?

hrobarts replied:
I've updated this parameter and the description - hopefully clearer now?

lauramurgatroyd commented on Wrappers/Python/test/test_DataProcessor.py:
```diff
@@ -3076,6 +3078,197 @@ def test_PaganinProcessor_2D(self):
+        processor = FluxNormaliser(roi=roi)
+        processor.set_input(self.data_cone)
+        data_norm = processor.get_output()
+        numpy.testing.assert_allclose(data_norm.array, self.data_cone.array)
```
is it because the dataset is simulated so theres the same mean flux?

lauramurgatroyd commented on Wrappers/Python/test/test_DataProcessor.py:
```diff
@@ -3076,6 +3078,197 @@ def test_PaganinProcessor_2D(self):
+        data_norm = processor.get_output()
+        numpy.testing.assert_allclose(data_norm.array, 5*self.data_cone.array)
+
+        # test roi with different data shapes
```
This test has different flux values in the roi at different projections which is great (the tests above do not have this). It might be good to point this out.

hrobarts replied:
I've added a comment

hrobarts commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,314 @@
+        equal to the number of angles in the dataset. If flux=None, calculate
+        flux from the roi, default is None.
+    
+    roi: dict (optional)
```
Hi Laura, I've added this text to the roi argument description and added a note similar to Binner. It's not exactly the same because we're not using step here.

hrobarts commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,314 @@
+                            axes.append(ax)
+                            self.roi.update({d:(0,dataset.get_dimension_size(d))})
+
+                    self.flux = numpy.mean(dataset.array[tuple(slc)], axis=tuple(axes))
```
I've moved this to a new function called `_calculate_flux()`

hrobarts commented on Wrappers/Python/test/test_DataProcessor.py:
```diff
@@ -3076,6 +3078,197 @@ def test_PaganinProcessor_2D(self):
+        numpy.testing.assert_allclose(data_norm.array, 0.5*self.data_cone.array)
+        
+        #Test flux array with no norm_value
+        processor = FluxNormaliser(flux=10*numpy.ones(self.data_cone.get_dimension_size('angle')))
```
I've added some new tests covering this

hrobarts commented:
Hi @lauramurgatroyd thank you for your review! I have made lots of updates
- More efficiently access data in process (following conversation with @gfardell )
- Error if horizontal and vertical are not in the last two dimensions of the data
- Change `norm_value` to `target` and allow it to be passed as a number or string `mean` or `first` which gets the target from the mean or first value of the flux array (e.g. this could be the mean value in the roi across all projections)
- Moved `_calculate_flux` and `_calculate_target` to separate functions out of `check_input`
- New tests to test this functionality

I have added the notebook where I created the examples above to CIL-Demos/misc in this PR https://github.com/TomographicImaging/CIL-Demos/pull/187

lauramurgatroyd reviewed (approved):
Looks great and nice examples in CIL-Demos misc.
As discussed this needs:
- [x]  an example of 'first' in the demo
- [ ] the demo being merged to main

hrobarts commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -29,133 +27,224 @@ class FluxNormaliser(Processor):
-                            
-                        # if the dimension is not in roi, average across the whole dimension
+                for d in ['horizontal', 'vertical']:
+                    if d in self.roi:
+                        # check indices are ints
+                        if not all(isinstance(i, int) for i in self.roi[d]):
+                            raise TypeError("roi values must be int, found {} and {}"
+                            .format(str(type(self.roi[d][0])), str(type(self.roi[d][1]))))
+                        # check indices are in range
+                        elif (self.roi[d][0] >= self.roi[d][1]) or (self.roi[d][0] < 0) or self.roi[d][1] > dataset.get_dimension_size(d):
+                            raise ValueError("roi values must be start > stop and between 0 and {}, found start={} and stop={} for direction '{}'"
+                            .format(str(dataset.get_dimension_size(d)), str(self.roi[d][0]), str(self.roi[d][1]), d ))
+                        # create slice
```
Hi Laura, this is a good point, I'll make an issue. I think the requirement for the ROI can vary between processors but there are definitely some common things.

hrobarts commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -29,133 +27,224 @@ class FluxNormaliser(Processor):
-            
-        return True
+        # check if flux array contains 0s
+        if 0 in self.flux:
```
Yes, I've added some tests.

gfardell commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,408 @@
+        
+        Parameters:
+        -----------
+        angle: float, optional
+            Angle to plot, default=None displays the data with the minimum
+            and maximum pixel values in the roi, otherwise the angle to display 
+            can be specified as a float and the closest angle will be displayed.
+            For 2D data, the roi is plotted on the sinogram.
```
Elsewhere I think we use 'angle' as the name of the axis, then the argument is the index of that projection in a list, not the value. It might be worth just discussing this further, not because it's wrong but it could just be a change of expected behaviour. Would it cause any issues if they specified an index?

It would also have to be careful about units - especially if we decide to convert things in the future and store in a defined way.

hrobarts replied:
Thanks Gemma, I've updated this now so it uses the index and I've also updated the preview_configuration plot so it shows the angle and the index

gfardell commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,408 @@
+        num_proj = int(data.array.size / proj_size)
+        
+        f = self.flux
+        for i in range(num_proj):
+            arr_proj = data.array.flat[i*proj_size:(i+1)*proj_size]
+            if len(self.flux.flat) > 1:
+                f = self.flux.flat[i] 
+            arr_proj *= self.target_value/f
+            out.array.flat[i*proj_size:(i+1)*proj_size] = arr_proj
```
This would be trivial to add a numba implementation, and here it would speed things up a lot. So I think we should look at that before we merge.

gfardell reviewed (changes requested):
This is looking great. I'd be happy to merge it in as it is. But I think it would be quick to add a numba implementation of the actual calculation. We can look at it together this week.

hrobarts commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,408 @@
+        in the image. The roi is specified as `{'axis_name1':(start,stop), 
+        'axis_name2':(start,stop)}`, where the key is the axis name 'vertical'
+        and/ or 'horizontal'. If an axis is not specified in the roi dictionary, 
+        the full range will be used.
```
Good point, updated now

gfardell commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,458 @@
+    in the roi dictionary
+    '''
+
+    def __init__(self, flux=None, roi=None, target='mean', accelerated=True):
```
maybe for a discussion, but do we need to expose `accelerated` here? As numba is a dependency I'd be happy to use it without giving an alternative route through the code.

gfardell commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,458 @@
+            out_flat[i*proj_size:(i+1)*proj_size] =  proj
+    else:
+        for i in numba.prange(num_proj):
+            proj = arr_flat[i*proj_size:(i+1)*proj_size]*target/flux[i]
```
This is multiplying the array by target, then dividing by flux. It needs some brackets to make sure the array is multiplied directly by a single value - this could be the cause of the timings you see for small datasets.

hrobarts replied:
Fixed now

gfardell commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,458 @@
+        ax1.set_xlabel(h)
+        ax1.set_ylabel(v)
+
+    @property
+    def num_threads(self):
+        '''Get the optional number of threads parameter to use for the accelerated version.
+
+        Defaults to the value set in the CIL multiprocessing module.'''
+        return cil_mp.NUM_THREADS if self._num_threads is None else self._num_threads
+
+    def set_num_threads(self, value):
+        '''Set the optional number of threads parameter to use for the accelerated version.
+
+        This is discarded if ``accelerated=False``.'''
+        self._num_threads = value
```
I'm not sure it's worth exposing this or not. in general we set the threads to half the number of cores as we don't check if hyperthreading is on or off. These operations are completely memory limited, not compute so more threads won't help over physical core numbers.

gfardell reviewed (changes requested):
Some small changes & discussion points. But it's looking good.

hrobarts commented:
Thank you for review @gfardell ! 
I think I've updated out so it can process in place. 
Here are some measurements of the execution time for the real FluxNormaliser class. This was on my cloud machine. I found this was pretty variable still. I can try on my windows machine too. Maybe this shows the loop isn't set up properly?

```
runtime = [0, 0, 0, 0, 0, 0]
flux = np.linspace(1, 1000, data.get_dimension_size('angle'), dtype=np.float32)
for i, threads in enumerate([0, 1, 5, 15, 30, 60]):
    
    if threads == 0:
        processor = FluxNormaliser(flux=flux, target=10, accelerated=False)
    else:
        processor = FluxNormaliser(flux=flux, target=10, accelerated=True)
        processor._num_threads = threads
    processor.set_input(data)
    data_norm = processor.get_output()
    t0 = time.time()
    for j in range(N):
        data_norm = processor.get_output()
    runtime[i] = (time.time()-t0)/N
    print(runtime[i])

x = ['parallel=False','N_threads=1','N_threads=5','N_threads=15','N_threads=30','N_threads=60']
plt.plot(x, runtime)
plt.grid()
plt.gca().tick_params(axis='x', labelrotation=45)
plt.ylabel('Execution time (seconds)')
plt.title('data.shape = {}'.format(str(data.shape)))
```
Using N=1000 for a small dataset:
![image](https://github.com/user-attachments/assets/b541cdb5-525f-4f28-9e60-6aeacdb43778)
And N=10 for a big dataset:
![image](https://github.com/user-attachments/assets/920a1185-841f-441d-bcf8-31a06011e22d)

hrobarts commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,458 @@
+        self._calculate_target()
+
+        data = self.get_input()
+        if out is None:
+            out = data.copy()
```
This is updated now

hrobarts commented on Wrappers/Python/cil/processors/FluxNormaliser.py:
```diff
@@ -0,0 +1,458 @@
+        norm = target/flux[0]
+        for i in numba.prange(num_proj):
+            proj = arr_flat[i*proj_size:(i+1)*proj_size]*norm
+            out_flat[i*proj_size:(i+1)*proj_size] =  proj
```
Updated now

hrobarts commented:
Compare different methods of indexing the data array with numba
- Test 1 ravel at each projection then scale whole projection
```
@numba.njit(parallel=True)
def numba_loop_1(flux, target, num_proj, proj_size, out):
        for i in numba.prange(num_proj):
            out_flat = out[i].ravel()
            for ij in range(proj_size):
                out_flat[ij] *= (target/flux[i])
```
- Test 2 flatten array then scale whole projection
```
@numba.njit(parallel=True)
def numba_loop_2(flux, target, num_proj, proj_size, out):
        out_flat = out.ravel()
        for i in numba.prange(num_proj):
            for ij in range(proj_size):
                out_flat[i*proj_size+ij] *= (target/flux[i])
```
- Test 3 index range of pixels
```
@numba.njit(parallel=True)
def numba_loop_3(flux, target, num_proj, proj_size, out):
        for i in numba.prange(num_proj):
            out_flat[i*proj_size:(i+1)*proj_size] *= (target/flux[i])
```
- Serial test 1 index range of pixels
```
def serial_loop1(flux, target, num_proj, proj_size, out):
        out_flat = out.ravel()
        for i in range(num_proj):
            out_flat[i*proj_size:(i+1)*proj_size] *= (target/flux[i])
```
- Serial test 2 use numpy multiply
```
def serial_loop2(flux, target, num_proj, proj_size, out):
    
    if len(flux) == 1:
        np.multiply(out, target/flux[0], out=out)
    else:
        for i in range(num_proj):
            out_flat = out[i].ravel()
            np.multiply(out_flat,  (target/flux[i]), out=out_flat)
```
- Serial test 3 same as test 1 but with parallel=False
```
@numba.njit(parallel=False)
def serial_loop3(flux, target, num_proj, proj_size, out):
        for i in numba.prange(num_proj):
            out_flat = out[i].ravel()
            for ij in range(proj_size):
                out_flat[ij] *= (target/flux[i])
```


- Tests with a flux array
![image](https://github.com/user-attachments/assets/bfa16247-9d6e-43c6-9a62-e082a8c8df00)
Speed up compared to the fastest serial method
```
         Pixels         Test 1         Test 2         Test 3
           1024            2.1           2.75           2.34
           2048           5.02           4.38           2.69
           3072           3.91           3.97           2.72
```
- Tests with a constant flux value
![image](https://github.com/user-attachments/assets/9bc66c44-3eb9-4226-9592-afc0482fbe3a)
Speed up compared to the fastest serial method
```
         Pixels         Test 1         Test 2         Test 3
           1024           1.43           1.64           1.36
           2048           2.81           2.88           2.01
           3072           3.92           5.32           2.81
```
- Tests measuring execution time for different numbers  of threads (choosing test2 and pixel size 3072), 
![image](https://github.com/user-attachments/assets/591cc322-9196-4a14-94a7-8b8f1ca9831e)

gfardell reviewed (approved):
Great job :) Run and tested via the linked notebook.
